### ISSUE TITLE
sharding jdbc, when insert config useGeneratedKeys , in one Transactional , the ID obtained by batch insertion is the first one! How to solve this problem

### ISSUE THREAD

**Issue by jiuye9999** [apache/shardingsphere#29245]
## Question

**For English only**, other languages will not accept.

Before asking a question, make sure you have:

- Googled your question.
- Searched open and closed [GitHub issues](https://github.com/apache/shardingsphere/issues).
- Read documentation: [ShardingSphere Doc](https://shardingsphere.apache.org/document/current/en/overview).

Please pay attention on issues you submitted, because we maybe need more details. 
If no response anymore and we cannot reproduce it on current information, we will **close it**.

**Comment by jiuye9999** [comment]
mysql sql: <insert id="insert" parameterType="com.example.shardingspheredemon.entity.BdfTeacher" useGeneratedKeys="true" keyProperty="id">
    insert into `bdf_teacher` (`teacher_name`, `class_id`)
    values (#{teacherName}, #{classId})
  </insert>

**Comment by bingo906** [comment]
I encountered the same issue in sharding-jdbc 4.1.1. Based on my analysis, it was because in org.apache.shardingsphere.shardingjdbc.jdbc.core.statement.ShardingPreparedStatement#prepare, only the lastId was retrieved. I'm not sure about the rationale behind this design. However, when I upgraded to sharding-jdbc 5, this issue was fixed. The relevant method involved is org.apache.shardingsphere.driver.jdbc.core.statement.ShardingSpherePreparedStatement#createExecutionContext.

**Comment by bingo906** [comment]
> mysql sql: insert into `bdf_teacher` (`teacher_name`, `class_id`) values (#{teacherName}, #{classId})

I encountered the same issue in sharding-jdbc 4.1.1. Based on my analysis, it was because in org.apache.shardingsphere.shardingjdbc.jdbc.core.statement.ShardingPreparedStatement#prepare, only the lastId was retrieved. I'm not sure about the rationale behind this design. However, when I upgraded to sharding-jdbc 5, this issue was fixed. The relevant method involved is org.apache.shardingsphere.driver.jdbc.core.statement.ShardingSpherePreparedStatement#createExecutionContext.
